### PR TITLE
updated attention mask value based on dtype.

### PR DESCRIPTION
**Pull request by kamalkraj**
# What does this PR do?

Fixes NaN output of `SelfAttention`/`MultiHeadDotProductAttention` layer with `dtype` float16

Fixes # (issue)
https://github.com/google/flax/issues/982

## Checklist
- [x] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other
      checks if that's the case).
- [ ] This change is discussed in a Github issue/
      [discussion](https://github.com/google/flax/discussions) (please add a
      link).
- [ ] The documentation and docstrings adhere to the
      [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [ ] This change includes necessary high-coverage tests.
      (No quality testing = no merge!)

### CODE REVIEW

File: flax/linen/attention.py [hunk 1]
```diff
@@ -287,10 +287,14 @@ def __call__(self,
...
       # attention mask in the form of attention bias
+      if self.dtype == jnp.float16:
+        large_compatible_negative = jnp.finfo(self.dtype).min
+      else:
+        large_compatible_negative = -1e10
       attention_bias = lax.select(
           mask > 0,
           jnp.full(mask.shape, 0.).astype(self.dtype),
-          jnp.full(mask.shape, -1e10).astype(self.dtype))
+          jnp.full(mask.shape, large_compatible_negative).astype(self.dtype))
```
Comment (levskaya) [inline flax/linen/attention.py:297]: Could we make this change more concise by just adding a `jnp.clip` inline here instead of introducing a new long-winded variable name? e.g.
`jnp.clip(-1e10, jnp.finfo(self.dtype).min)`

(I'm just sensitive to trying not to let this layer code grow too much or add new names - there's already a lot of detail for a new reader to keep track of.)

**Review by levskaya — CHANGES_REQUESTED** [review]
Thanks so much for catching this!   Just a style request below.

File: flax/linen/attention.py (hunk 1 above)
Comment (kamalkraj) [reply flax/linen/attention.py]: done.

**Comment by levskaya** [comment]
@kamalkraj - oof, please forgive us for some confusion:  

After we looked at this correction we realized there was a much more comprehensive PR addressing mixed-precision training on WMT from several weeks ago that we'd forgotten to merge.  That PR is #1414 which further introduces dynamic loss scaling to improve training stability with `float16`.  Thank you so much for spotting this error and prodding us to get these GPU-compatibility changes in for this example.  

Since this resolves the original issue as well, I'll close this PR for now but as always feel free to open new ones if we're still missing anything.  Thanks again.
